www/js/splash/notifScheduler.js

```diff
@@ -0,0 +1,190 @@
+            for (const d of notifDays) {
+                const date = moment(dayZeroDate).add(d, 'days').format('YYYY-MM-DD')
+                const notifTime = moment(date+' '+timeOfDay, 'YYYY-MM-DD HH:mm');
+                notifTimes.push(notifTime);
+            }
+        }
+        return notifTimes;
+    }
+
+    // returns true if all expected times are already scheduled
+    const areAlreadyScheduled = (notifs, expectedTimes) => {
+        for (const t of expectedTimes) {
+            if (!notifs.some((n) => moment(n.at).isSame(t))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    const setUpActions = () => {
+        const action = {
+            id: 'action',
+            title: 'Change Time',
+            launch: true
+        };
+        return new Promise((rs) => {
+            cordova.plugins.notification.local.addActions('reminder-actions', [action], rs);
+        });
+    }
+
+    function debugGetScheduled(prefix) {
+        cordova.plugins.notification.local.getScheduled((notifs) => {
+            if (!notifs?.length)
+                return Logger.log(`${prefix}, there are no scheduled notifications`);
+            const time = moment(notifs?.[0].trigger.at).format('HH:mm');
+            Logger.log(`${prefix}, there are ${notifs.length} scheduled notifications at ${time}`);
+        });
+    }
+
+    // schedules the notifications using the cordova plugin
+    const scheduleNotifs = (scheme, notifTimes) => {
+        return new Promise((rs) => {
+            isScheduling = true;
+            const nots = notifTimes.map((n) => {
+                const nDate = n.toDate();
+                const seconds = nDate.getTime() / 1000;
+                return {
+                    id: seconds,
+                    title: scheme.title,
+                    text: scheme.text,
+                    trigger: {at: nDate},
+                    actions: 'reminder-actions',
+                    data: {
+                        action: {
+                            redirectTo: 'root.main.control',
+                            redirectParams: {
+                                openTimeOfDayPicker: true
+                            }
+                        }
+                    }
+                }
+            });
+            cordova.plugins.notification.local.cancelAll(() => {
+                debugGetScheduled("After cancelling");
+                cordova.plugins.notification.local.schedule(nots, () => {
+                    debugGetScheduled("After scheduling");
+                    isScheduling = false;
+                    rs();
+                });
+            });
+        });
+    }
+
+    // determines when notifications are needed, and schedules them if not already scheduled
+    const update = async () => {
+        const { reminder_assignment,
+                reminder_join_date,
+                reminder_time_of_day} = await scheduler.getReminderPrefs();
```

Comment:
For my own edification, the call chain for assigning people to groups is:
- when the app is launched (`ionicPlatform` is run), we call `update`
- update calls `getReminderPrefs()`
- which calls `initReminderPrefs()` since there are no stored preferences
   - which randomly assigns the user to one of the groups

I think this should work for the other two cases as well, but it would be good if you could list out the flow for:
- launching the app (common case)?
- changing the preferred time?

Reply:
When launching the app any subsequent time after onboarding/assignment:
- once `ionicPlatform.ready()`, we call `update()`
- which calls `getReminderPrefs()`
- since the reminder prefs have already been initialized, the stored prefs are retrieved and used
- the scheduled notifications are checked against the assignment. If out of sync (in this case, no), the schedule is recomputed

When changing the "Time of Day" preference:
- `setReminderPrefs()` is called
- the new "Time of Day" preference is stored
- `update()` is called
- the scheduled notifications are checked against the assignment. If out of sync (in this case, yes), the schedule is recomputed